### PR TITLE
ESP32-S2: Add support for Main System Watchdog Timers

### PR DESCRIPTION
**Pull request by gustavonihei**
## Summary
This MR intends to add support for the Main System Watchdog Timers 0 and 1 of **ESP32-S2**.

Support for RTC Watchdog Timer is currently in place, but not yet functional due to not yet implemented RTC driver.

## Impact
New feature for **ESP32-S2** chip. 

## Testing
Validated with the Watchdog example application from `nuttx-apps`.
image

### CODE REVIEW

File: arch/xtensa/src/esp32s2/esp32s2_wdt_lowerhalf.c [hunk 1]
```diff
@@ -0,0 +1,740 @@
...
+  leave_critical_section(flags);
+  ESP32S2_WDT_LOCK(priv->wdt);
+  return oldhandler;
+}
+
+/* Interrupt handling *******************************************************/
+
+static int wdt_handler(int irq, void *context, void *arg)
+{
+  struct esp32s2_wdt_lowerhalf_s *priv = arg;
```
Comment (pkarashchenko) [inline arch/xtensa/src/esp32s2/esp32s2_wdt_lowerhalf.c:609]:
```suggestion
  struct esp32s2_wdt_lowerhalf_s *priv = (struct esp32s2_wdt_lowerhalf_s *)arg;
```

**Comment by gustavonihei** [comment]
@pkarashchenko I've addressed the raised issues. Please, take another look. Thanks!

File: arch/xtensa/src/esp32s2/esp32s2_wdt_lowerhalf.c [hunk 2]
```diff
@@ -0,0 +1,741 @@
...
+
+static const struct watchdog_ops_s g_esp32s2_wdg_ops =
+{
+  .start      = wdt_lh_start,
+  .stop       = wdt_lh_stop,
+  .keepalive  = wdt_lh_keepalive,
+  .getstatus  = wdt_lh_getstatus,
+  .settimeout = wdt_lh_settimeout,
+  .capture    = wdt_lh_capture,
+  .ioctl      = NULL,
```
Comment (pkarashchenko) [inline arch/xtensa/src/esp32s2/esp32s2_wdt_lowerhalf.c:132]:
```suggestion
  .ioctl      = NULL
```

File: arch/xtensa/src/esp32s2/esp32s2_wdt_lowerhalf.c [hunk 3]
```diff
@@ -0,0 +1,741 @@
...
+};
+
+#ifdef CONFIG_ESP32S2_MWDT0
+/* MWDT0 lower-half */
+
+static struct esp32s2_wdt_lowerhalf_s g_esp32s2_mwdt0_lowerhalf =
+{
+  .ops = &g_esp32s2_wdg_ops,
```
Comment (pkarashchenko) [inline arch/xtensa/src/esp32s2/esp32s2_wdt_lowerhalf.c:140]:
```suggestion
  .ops = &g_esp32s2_wdg_ops
```

File: arch/xtensa/src/esp32s2/esp32s2_wdt_lowerhalf.c [hunk 4]
```diff
@@ -0,0 +1,741 @@
...
+};
+#endif
+
+#ifdef CONFIG_ESP32S2_MWDT1
+/* MWDT1 lower-half */
+
+static struct esp32s2_wdt_lowerhalf_s g_esp32s2_mwdt1_lowerhalf =
+{
+  .ops = &g_esp32s2_wdg_ops,
```
Comment (pkarashchenko) [inline arch/xtensa/src/esp32s2/esp32s2_wdt_lowerhalf.c:149]:
```suggestion
  .ops = &g_esp32s2_wdg_ops
```

File: arch/xtensa/src/esp32s2/esp32s2_wdt_lowerhalf.c [hunk 5]
```diff
@@ -0,0 +1,741 @@
...
+};
+#endif
+
+#ifdef CONFIG_ESP32S2_RWDT
+/* RWDT lower-half */
+
+static struct esp32s2_wdt_lowerhalf_s g_esp32s2_rwdt_lowerhalf =
+{
+  .ops = &g_esp32s2_wdg_ops,
```
Comment (pkarashchenko) [inline arch/xtensa/src/esp32s2/esp32s2_wdt_lowerhalf.c:158]:
```suggestion
  .ops = &g_esp32s2_wdg_ops
```

File: arch/xtensa/src/esp32s2/esp32s2_wdt_lowerhalf.c [hunk 6]
```diff
@@ -0,0 +1,741 @@
...
+  wdinfo("Entry: wdt_lh_start\n");
+
+  DEBUGASSERT(priv);
+
+  if (priv->started)
+    {
+      /* Return EBUSY to indicate that the timer was already running */
+
+      ret = -EBUSY;
+      goto errout;
```
Comment (pkarashchenko) [inline arch/xtensa/src/esp32s2/esp32s2_wdt_lowerhalf.c:194]:
```suggestion
```

File: arch/xtensa/src/esp32s2/esp32s2_wdt_lowerhalf.c [hunk 7]
```diff
@@ -0,0 +1,741 @@
...
+            if (priv->peripheral == TIMER)
+              {
+                ESP32S2_WDT_STG_CONF(priv->wdt, ESP32S2_WDT_STAGE0,
+                                     ESP32S2_WDT_STAGE_ACTION_RESET_SYSTEM);
+              }
+            else
+              {
+                ESP32S2_WDT_STG_CONF(priv->wdt, ESP32S2_WDT_STAGE0,
+                                     ESP32S2_WDT_STAGE_ACTION_RESET_RTC);
+              }
```
Comment (pkarashchenko) [inline arch/xtensa/src/esp32s2/esp32s2_wdt_lowerhalf.c:224]: too many spaces

File: arch/xtensa/src/esp32s2/esp32s2_wdt_lowerhalf.c [hunk 8]
```diff
@@ -0,0 +1,741 @@
...
+      priv->lastreset = clock_systime_ticks();
+      ESP32S2_WDT_START(priv->wdt);
+      leave_critical_section(flags);
+
+      /* Lock it again */
+
+      ESP32S2_WDT_LOCK(priv->wdt);
+    }
+
+  errout:
```
Comment (pkarashchenko) [inline arch/xtensa/src/esp32s2/esp32s2_wdt_lowerhalf.c:254]:
```suggestion
```

File: arch/xtensa/src/esp32s2/esp32s2_wdt_lowerhalf.c (hunk 7 above)
Comment (pkarashchenko) [reply arch/xtensa/src/esp32s2/esp32s2_wdt_lowerhalf.c:223]: I'm not sure why style check does not catch this

Comment (gustavonihei) [reply arch/xtensa/src/esp32s2/esp32s2_wdt_lowerhalf.c:223]: Sorry, this one I couldn't understand. Are you referring to lines 218 and 223? They are compliant to the [Line Wrapping](https://nuttx.apache.org/docs/latest/contributing/coding_style.html#lines) item of the Coding Standard.

Comment (pkarashchenko) [reply arch/xtensa/src/esp32s2/esp32s2_wdt_lowerhalf.c:223]:
```suggestion
          /* Then configure it to reset on wdt expiration */

          if (priv->peripheral == TIMER)
            {
              ESP32S2_WDT_STG_CONF(priv->wdt, ESP32S2_WDT_STAGE0,
                                   ESP32S2_WDT_STAGE_ACTION_RESET_SYSTEM);
            }
          else
            {
              ESP32S2_WDT_STG_CONF(priv->wdt, ESP32S2_WDT_STAGE0,
                                   ESP32S2_WDT_STAGE_ACTION_RESET_RTC);
            }
```

Comment (gustavonihei) [reply arch/xtensa/src/esp32s2/esp32s2_wdt_lowerhalf.c:223]: Oh, sorry, now I found the issue on the whole block indented twice. Thanks!

File: arch/xtensa/src/esp32s2/esp32s2_wdt_lowerhalf.c [hunk 9]
```diff
@@ -0,0 +1,739 @@
...
+  if (lower->upper == NULL)
+    {
+      /* The actual cause of the failure may have been a failure to allocate
+       * perhaps a failure to register the watchdog driver (such as if the
+       * 'devpath' were not unique).  We know here but we return EEXIST to
+       * indicate the failure (implying the non-unique devpath).
+       */
+
+      ret = -EEXIST;
+      goto errout;
```
Comment (pkarashchenko) [inline arch/xtensa/src/esp32s2/esp32s2_wdt_lowerhalf.c:734]:
```suggestion
```

**Review by pkarashchenko — APPROVED** [review]
LGTM!

**Pull request merged**